platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/TaskList.java

```diff
@@ -0,0 +1,215 @@
+   * unique identifier for each task to prevent duplication.
+   */
+  private final Map<Operation, Map<String, Predicate<ProfileMigrationReport>>> groups =
+      new EnumMap<>(Operation.class);
+
+  /**
+   * Holds attempts counters for each operations.
+   *
+   * <p>The counters stored here will start at the defined retry counter and once they reached 0,
+   * the maximum attempts on the corresponding operation will have been reached and we shall stop
+   * executing a particular group/operation of tasks for a give type of entities (i.e. apps,
+   * features, or bundles).
+   */
+  private final Map<Operation, AtomicInteger> attemptsLeft = new EnumMap<>(Operation.class);
+
+  private final String type;
+
+  private final ProfileMigrationReport report;
+
+  /**
+   * Constructs a new task list for the given type of objecs.
+   *
+   * @param type the type of object managed by this task list
+   * @param report the report where to report that tasks are being recorded
+   * @throws IllegalArgumentException if <code>type</code> or <code>report</code> is <code>null
+   *     </code>
+   */
+  public TaskList(String type, ProfileMigrationReport report) {
+    Validate.notNull(type, "invalid null task type");
+    Validate.notNull(report, "invalid null report");
+    this.type = type;
+    this.report = report;
+    for (final Operation op : Operation.values()) {
+      attemptsLeft.put(op, new AtomicInteger(ProfileMigratable.RETRY_COUNT));
+    }
+  }
+
+  /**
+   * Adds a task to be executed for a specific operation. The task will be provided with a profile
+   * migration report where to record messages. It should return <code>true</code> if the task was
+   * successful or <code>false</code> if not.
+   *
+   * @param op the operation for which to record the task
+   * @param id a unique identifier for the task (e.g. feature name)
+   * @param task the task to add for later execution
+   * @throws IllegalArgumentException if <code>op</code>, <code>id</code>, or <code>task</code> is
+   *     <code>null</code>
+   */
+  public void add(Operation op, String id, Predicate<ProfileMigrationReport> task) {
+    Validate.notNull(op, "invalid null operation");
+    Validate.notNull(id, "invalid null task id");
+    Validate.notNull(task, "invalid null task");
+    LOGGER.debug("Recording {} task for {} '{}'", op, type, id);
+    report.recordTask();
+    groups
+        .computeIfAbsent(op, o -> new LinkedHashMap<>()) // preserve order of execution
+        .put(id, task);
+  }
+
+  /**
+   * Increases the number of attempts left for the specified operation
+   *
+   * @param op the operation to increase the number of attempts for
+   * @throws IllegalArgumentException if <code>op</code> is <code>null</code>
+   */
+  public void increaseAttemptsFor(Operation op) {
+    Validate.notNull(op, "invalid null operation");
+    // increase the number of attempts left for the given operation and make sure we have at lease 1
```

Comment:
✏️  Typo: at _lease_. Not sure that this comment really adds anything actually.

Reply:
removed and method is going away in next PR